File: Cpp solutions/20 valid parentheses.cpp
// https://leetcode.com/problems/valid-parentheses
// TC: O(n) SC: O(1)
class Solution {
public:
    bool isValid(string s) {
        bool isParenthesesValid = true;
        int index = 0;
        stack<char> parentheses;
        while (isParenthesesValid == true && index < s.size()){
            if (s[index] == '(' || s[index] == '[' || s[index] == '{'){
                parentheses.push(s[index]);
            }
            else if (s[index] == ')'){
                if (parentheses.empty() != true && parentheses.top() == '('){
                    parentheses.pop();
                }
                else{
                    isParenthesesValid = false;
                }
            }
            else if (s[index] == ']'){
                if (parentheses.empty() != true && parentheses.top() == '['){
                    parentheses.pop();
                }
                else{
                    isParenthesesValid = false;
                }
            }
            else if (s[index] == '}'){
                if (parentheses.empty() != true && parentheses.top() == '{'){
                    parentheses.pop();
                }
                else{
                    isParenthesesValid = false;
                }
            }
            else{
                isParenthesesValid = false;
            }
            index++;
        }
        if (parentheses.empty() != true){
            isParenthesesValid = false;
        }
        return isParenthesesValid;
    }
};

File: Python solutions/20 valid parentheses .py
# https://leetcode.com/problems/valid-parentheses
# TC: o(n) SC: o(n)
# Stack
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []

        for c in s:
            if c == '(' or c == '[' or c == '{':
                stack.append(c)
            elif stack:
                par = stack.pop()
                if (c == ')' and par != '(') or (c == ']' and par != '[') or (c == '}' and par != '{'):
                    return False
            else:
                return False
        return True if len(stack) == 0 else False

